osquery/tables/applications/darwin/browser_firefox.cpp

```diff
@@ -0,0 +1,122 @@
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/property_tree/json_parser.hpp>
+
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+/// Each home directory will include custom extensions.
+#define kFirefoxPath "/Library/Application Support/Firefox/Profiles/"
+#define kFirefoxExtensionsFile "/extensions.json"
+
+/// Not parsed, but may be helpful later.
+#define kFirefoxAddonsFile "/addons.json"
+#define kFirefoxWebappsFile "/webapps/webapps.json"
+
+const std::map<std::string, std::string> kFirefoxAddonKeys = {
+    {"defaultLocale.name", "name"},
+    {"id", "identifier"},
+    {"type", "type"},
+    {"version", "version"},
+    {"defaultLocale.creator", "creator"},
+    {"defaultLocale.description", "description"},
+    {"sourceURI", "source_url"},
+    {"visible", "visible"},
+    {"active", "active"},
+    {"applyBackgroundUpdates", "autoupdate"},
+    {"hasBinaryComponents", "native"},
+    {"location", "location"},
+    {"descriptor", "path"},
+};
+
+void genFirefoxAddonsFromExtensions(const std::string& path,
+                                    QueryData& results) {
+  std::string json_data;
+  if (!readFile(path + kFirefoxExtensionsFile, json_data).ok()) {
+    VLOG(1) << "Could not read file: " << path + kFirefoxExtensionsFile;
+    return;
+  }
+
+  // Read the extensions data into a JSON blob, then property tree.
+  pt::ptree tree;
+  std::stringstream json_stream;
+  json_stream << json_data;
+  try {
+    pt::read_json(json_stream, tree);
+  } catch (const pt::json_parser::json_parser_error& e) {
+    VLOG(1) << "Could not parse JSON from: " << path + kFirefoxExtensionsFile;
+    return;
+  }
+
+  for (const auto& addon : tree.get_child("addons")) {
+    Row r;
+    // Most of the keys are in the top-level JSON dictionary.
+    for (const auto& it : kFirefoxAddonKeys) {
+      try {
+        r[it.second] = addon.second.get<std::string>(it.first);
+      } catch (const pt::ptree_error& e) {
+        r[it.second] = "";
+      }
+
+      // Convert bool-types to an integer.
+      if (r[it.second] == "true" || r[it.second] == "YES" ||
+          r[it.first] == "Yes") {
+        r[it.second] = INTEGER(1);
+      } else if (r[it.second] == "false" || r[it.second] == "NO" ||
+                 r[it.second] == "No") {
+        r[it.second] = INTEGER(0);
+      }
+    }
+
+    // There are several ways to disabled the addon, check each.
+    try {
+      if (addon.second.get<std::string>("softDisable") == "true" ||
+          addon.second.get<std::string>("appDisabled") == "true" ||
+          addon.second.get<std::string>("userDisabled") == "true") {
+        r["disabled"] = INTEGER(1);
+      } else {
+        r["disabled"] = INTEGER(0);
+      }
+    } catch (const pt::ptree_error& e) {
+      r["disabled"] = "0";
```

Comment:
It's possible for any one of these keys to not exist, resulting in the catch. A better approach would be to check for each individually. Only the specific value == true for any of these keys means disabled, missing or otherwise means enabled.